Support key deletion in Data Protection

amcasey opened the pull request:
Add the ability to delete keys to support #52916.  The approach is more complicated than you'd expect to account for the fact that `IXmlRepository` offers only two (possibly slow) operations: enumerate and add - there's no random access.

Deleting keys remains discouraged.

Fixes #53880

amcasey committed:
Add key deletion APIs to IKeyManager

amcasey committed:
Add element removal APIs to IXmlRepository

amcasey committed:
Implement IXmlRepository deletion for EF, FS, and registry

amcasey committed:
Make API manifest per-framework

amcasey committed:
Implement missing logging methods

amcasey committed:
Update IKeyManager API to parallel IXmlRepository

amcasey committed:
Pass entire list of XElements to RemoveElements callback

amcasey committed:
Started implementing key manager deletion and hit roadblock

amcasey committed:
Return success from IXmlRepository.RemoveElements

amcasey committed:
Finish implementing key manager deletion

amcasey committed:
Add more logging

amcasey committed:
Consume unsafeIncludeUnexpired

amcasey committed:
Change IXmlRepository.RemoveElements to use a mutating Action

amcasey committed:
Tidy up diff

amcasey commented on src/DataProtection/DataProtection/src/LoggingExtensions.cs:
```diff
@@ -238,15 +238,36 @@ private static bool IsLogLevelEnabledCore([NotNullWhen(true)] ILogger? logger, L
     [LoggerMessage(60, LogLevel.Warning, "Storing keys in a directory '{path}' that may not be persisted outside of the container. Protected data will be unavailable when container is destroyed. For more information go to https://aka.ms/aspnet/dataprotectionwarning", EventName = "UsingEphemeralFileSystemLocationInContainer")]
     public static partial void UsingEphemeralFileSystemLocationInContainer(this ILogger logger, string path);
 
-    [LoggerMessage(61, LogLevel.Trace, "Ignoring configuration '{PropertyName}' for options instance '{OptionsName}'", EventName = "IgnoringReadOnlyConfigurationForNonDefaultOptions")]
```
I guess 60 was missed, so it appears after 66, which seems to have confused whoever added the following items.  Can I update these, given that they're wrong now?

amcasey commented:
I had to move to per-target-platform API manifests because DIM isn't available on framework.

amcasey commented:
I'll add tests once the API is reviewed, but the shape has changed several times already.

danmoseley commented on src/DataProtection/DataProtection/src/Repositories/RegistryXmlRepository.cs:
```diff
@@ -155,4 +155,63 @@ private void StoreElementCore(XElement element, string valueName)
+        return allSucceeded;
+    }
+
+    private sealed class DeletableElement : IDeletableElement
```
I believe it's `Deleteable` not `Deletable` ?

amcasey replied:
I believe both are acceptable but no-e is more common, at least in the US.

amcasey committed:
Fix merge

amcasey commented:
Force push is a rebase - I'll address API Review feedback separately.

amcasey committed:
Remove unsafeIncludeUnexpired

amcasey committed:
Switch from ShouldDelete to DeletionOrder

amcasey commented:
I also prepared a version where the key manager gets to return an ordered list of `IDeletableElements` to delete, but I was unhappy with the fact that the implementation could return duplicate elements or elements that were not in the list it received.  These are both solvable problems, but every repository implementation would need these safeguards, which I found less appealing than having every repository sort.

amcasey commented:
Force push is a trivial rebase on `main`.

amcasey committed:
Improve comments

amcasey committed:
Implement deletion in EphemeralXmlRepository

amcasey committed:
Add comment requested by API Review

amcasey committed:
Stub out RedisXmlRepository implementation for API Review

amcasey commented:
TODO: Redis impl
TODO: add tests

amcasey committed:
Split out new interfaces so we don't need ifdefs

amcasey commented on src/DataProtection/DataProtection/src/KeyManagement/IKeyManagerWithDeletion.cs:
```diff
@@ -0,0 +1,51 @@
+    /// Deletion is stronger than revocation.  A revoked key is retained and can even be (forcefully) applied.
+    /// A deleted key is indistinguishable from a key that never existed.
+    /// </remarks>
+    bool CanDeleteKeys { get; }
```
@halter73 I kept this because deletion will fail if the underlying repository doesn't support it and I thought it made sense to provide a way to query for that, rather than just trying and failing.

amcasey replied:
I was going to send mail with the final shape, but discussing it Monday might be quicker.

amcasey committed:
Actually stop iterating if some key deletion fails

amcasey committed:
Add FileSystemXmlRepository tests

halter73 commented on src/DataProtection/DataProtection/src/KeyManagement/IKeyManagerWithDeletion.cs:
```diff
@@ -0,0 +1,51 @@
+/// <remarks>
+/// Instantiations of this interface are expected to be thread-safe.
+/// </remarks>
+public interface IKeyManagerWithDeletion : IKeyManager
```
I thought we approved `IDeletableKeyManager` and `IDeletableXmlRepository` to correspond with `IDeletableElement`.

amcasey replied:
I don't remember picking names, but those seem fine.

amcasey committed:
Add RegistryXmlRepository tests

amcasey committed:
Rename new interfaces

amcasey commented on src/DataProtection/DataProtection/src/Repositories/FileSystemXmlRepository.cs:
```diff
@@ -16,7 +17,7 @@ namespace Microsoft.AspNetCore.DataProtection.Repositories;
 /// <summary>
 /// An XML repository backed by a file system.
 /// </summary>
-public class FileSystemXmlRepository : IXmlRepository
```
@halter73 Do I need to keep implementing the old interface as well for back compat?

halter73 replied:
What you have is fine.

> It is not breaking when you added the implementation of an interface which derives from the removed interface. For example, you removed `IDisposable`, but implemented `IComponent`, which derives from `IDisposable`.

[breaking-change-rules.md](https://github.com/dotnet/runtime/blob/856a0a6f5e991fd18e125bcfc88741d919688fbf/docs/coding-guidelines/breaking-change-rules.md#L162)

amcasey committed:
Defer EF and Redis to a subsequent PR

amcasey committed:
Add EphemeralXmlRepository tests

amcasey committed:
Revert EF logging

amcasey committed:
Update copy-pasta doc comments

amcasey committed:
Add XmlKeyManager tests

amcasey commented:
/azp run

halter73 commented on src/DataProtection/DataProtection/src/Repositories/EphemeralXmlRepository.cs:
```diff
@@ -54,4 +54,59 @@ public void StoreElement(XElement element, string friendlyName)
+
+        var deletableElements = new List<DeletableElement>();
+
+        foreach (var storedElement in _storedElements)
```
This needs to be inside `lock (_storedElements)` or risk an `InvalidOperationException` for enumerating the collection during modification.

amcasey replied:
Done, though I still think implemeting deletion on the ephemeral repository is silly.

halter73 commented on src/DataProtection/DataProtection/src/Repositories/EphemeralXmlRepository.cs:
```diff
@@ -54,4 +54,59 @@ public void StoreElement(XElement element, string friendlyName)
+        {
+            foreach (var deletableElement in elementsToDelete)
+            {
+                var storedElement = deletableElement.StoredElement;
+                var index = _storedElements.FindIndex(e => ReferenceEquals(e, storedElement));
+                if (index >= 0) // Might not find it if the collection has changed since we started.
+                {
+                    // It would be more efficient to remove the larger indices first, but deletion order
+                    // is important for correctness.
+                    _storedElements.RemoveAt(index);
+                }
```
Nit:

```suggestion
                // It would be more efficient to remove the larger indices first, but deletion order
                // is important for correctness.
                _storedElements.Remove(deletableElement.StoredElement);
```

This should still be fine even if we release and reacquire the lock allowing the element to be removed. `Remove` will just return false.

amcasey replied:
Remove uses `object.Equals` and I think we need `object.ReferenceEquals`.

halter73 reviewed (commented):
@Tratcher have you taken a look at this recently? Does it meet your needs?

halter73 commented on src/DataProtection/DataProtection/src/LoggingExtensions.cs:
```diff
@@ -255,4 +255,25 @@ private static bool IsLogLevelEnabledCore([NotNullWhen(true)] ILogger? logger, L
+    public static partial void DeletingFile(this ILogger logger, string fileName);
+
+    [LoggerMessage(75, LogLevel.Information, "Failed to delete file '{FileName}'.  Not attempting further deletions.", EventName = "FailedToDeleteFile")]
+    public static partial void FailedToDeleteFile(this ILogger logger, string fileName, Exception exception);
```
Why is this log information-level while `FailedToRemoveDataFromRegistryKeyValue` is error-level? Should it be error-level for both?

amcasey replied:
Looks like a copy-paste error though, in practice, neither level will actually be printed unless `--verbose` is passed.

amcasey commented on src/DataProtection/DataProtection/src/LoggingExtensions.cs:
```diff
@@ -255,4 +255,25 @@ private static bool IsLogLevelEnabledCore([NotNullWhen(true)] ILogger? logger, L
+    [LoggerMessage(74, LogLevel.Information, "Deleting file '{FileName}'.", EventName = "DeletingFile")]
+    public static partial void DeletingFile(this ILogger logger, string fileName);
+
+    [LoggerMessage(75, LogLevel.Information, "Failed to delete file '{FileName}'.  Not attempting further deletions.", EventName = "FailedToDeleteFile")]
```
```suggestion
    [LoggerMessage(75, LogLevel.Error, "Failed to delete file '{FileName}'.  Not attempting further deletions.", EventName = "FailedToDeleteFile")]
```

amcasey committed:
Correct log level

amcasey committed:
Enumerate _storedElements under lock

Tratcher commented on src/DataProtection/DataProtection/src/Repositories/IDeletableXmlRepository.cs:
```diff
@@ -0,0 +1,27 @@
+    /// <returns>
+    /// True if all deletions succeeded.
+    /// </returns>
+    bool RemoveElements(Action<IReadOnlyCollection<IDeletableElement>> chooseElements);
```
Inconsistent naming Remove vs Delete?

amcasey replied:
Do you remember why we did this, @halter73?  I'm pretty sure it was intentional.  If not, are you okay with changing it to `DeleteElements`, etc?

amcasey replied:
My guess is that `IXmlRepository` is collection-like and collections usually use "Remove".

amcasey replied:
I don't mind it being inconsistent with `IKeyManager`, but I do mind it being inconsistent with `IDeletableXmlRepository`.

amcasey replied:
Renamed to DeleteElements.

Tratcher commented on src/DataProtection/DataProtection/src/Repositories/RegistryXmlRepository.cs:
```diff
@@ -155,4 +155,62 @@ private void StoreElementCore(XElement element, string valueName)
+            catch (Exception ex)
+            {
+                _logger.FailedToRemoveDataFromRegistryKeyValue(RegistryKey, valueName, ex);
+                return false;
```
```suggestion
                // Stop processing deletions to avoid deleting a revocation entry for a key that we failed to delete.
                return false;
```

amcasey commented on src/DataProtection/DataProtection/src/KeyManagement/IDeletableKeyManager.cs:
```diff
@@ -0,0 +1,48 @@
+    /// <summary>
+    /// Deletes keys matching a predicate.
+    ///
+    /// Use with caution as deleting active keys will normally cause data loss.
```
It will unless you do something special.

Tratcher replied:
🤷 It's only data loss if you have something to lose.

amcasey replied:
It's specifically about active keys, so I think it's true in most cases.

amcasey commented on src/DataProtection/DataProtection/src/LoggingExtensions.cs:
```diff
@@ -255,4 +255,25 @@ private static bool IsLogLevelEnabledCore([NotNullWhen(true)] ILogger? logger, L
     [LoggerMessage(73, LogLevel.Debug, "Key {KeyId:B} method {MethodName} failed. Retrying.", EventName = "RetryingMethodOfKeyAfterFailure")]
     public static partial void RetryingMethodOfKeyAfterFailure(this ILogger logger, Guid keyId, string methodName, Exception exception);
+	
+    [LoggerMessage(74, LogLevel.Information, "Deleting file '{FileName}'.", EventName = "DeletingFile")]
```
TBH, I let copilot scaffold a lot of these. 😆

amcasey replied:
Nvm, I was thinking of a different PR.  That's probably just a copy-paste error.  Changed to `Debug`.

amcasey committed:
Rename RemoveElements to DeleteElements for consistency

amcasey committed:
Address PR feedback

amcasey committed:
Add clarifying comments.

Co-authored-by: Chris Ross <Tratcher@Outlook.com>

amcasey commented:
I'm definitely a little curious how my new tests could pass only on Windows.  Investigating.

amcasey committed:
Rename tests to match updated APIs

amcasey committed:
Make registry tests conditional on registry availability

amcasey committed:
Skip DeleteElementsWithFailure on Linux

The only way I could find to make FileSystemInfo.Delete throw was to remove write permission to the parent directory and we don't have enough hooks to do and undo that around a single file deletion attempt (i.e. it's all or none).

amcasey committed:
Also skip on mac

donnyv commented:
I found this while building a custom repository for Mongodb. 

One idea that may be to late, since you look like you are pretty far along. You could simplify this by just adding a DeleteElement(string keyid) to IXmlRepository. Then let devs developing their own repositories to create the delete code. DeleteElement() method could be fired by Data Protection using a TTL. A TTL could be added to every key automatically when its created. Since the minimum expiration that can be set to a Key is 7 days. Data Protection could check the TTL every 7 days and fire deletes for all the expired keys. TTL date could be (expired date + time to live length) = TTL. Time to live length could be a setting with a default of (expired date * 2). I would make it so that devs would have to turn this feature on too.

amcasey commented:
@donnyv The reason we didn't add `DeleteElement(string keyid)` is that random access is not presently part of the `IXmlRepository` contract and a number of existing implementations (e.g. file system) couldn't implement it efficiently.

Having said that, it was certainly our intention that `IXmlRepository` implementers would be able to provide their own implementations.  Are you finding that not to be the case?

Unless we hear a lot of demand (feel free to open an issue to collect feedback), I don't think we're likely to add an automated TTL mechanism.  Generally speaking, we recommend keys never be deleted, so we want doing so to be a very explicit action on the app author's part.
